File: backend/prayer_times.py
"""
Seydişehir namaz vakitleri hesaplayıcı.

Kaynak formüller: https://www.islamicfinder.org/prayer-times/calculation-methods/
Kullanılan yöntem: Diyanet İşleri Başkanlığı (Fajr 18°, Isha 17°, Asr Standard)
"""
import math
import datetime


LAT = 37.4167
LON = 31.8500
TZ_OFFSET = 3  # UTC+3 (Türkiye)

# Diyanet açıları
FAJR_ANGLE = 18.0
ISHA_ANGLE = 17.0


def _deg(x: float) -> float:
    return math.degrees(x)


def _rad(x: float) -> float:
    return math.radians(x)


def _fix_angle(a: float) -> float:
    a = a - 360.0 * math.floor(a / 360.0)
    return a if a >= 0 else a + 360.0


def _sun_position(jd: float) -> tuple[float, float]:
    """Julian day → (declination_deg, equation_of_time_hours)"""
    d = jd - 2451545.0
    g = _fix_angle(357.529 + 0.98560028 * d)
    q = _fix_angle(280.459 + 0.98564736 * d)
    L = _fix_angle(q + 1.9148 * math.sin(_rad(g)) + 0.0200 * math.sin(_rad(2 * g)))
    e = 23.439 - 0.0000004 * d
    RA = _deg(math.atan2(math.cos(_rad(e)) * math.sin(_rad(L)), math.cos(_rad(L)))) / 15.0
    dec = _deg(math.asin(math.sin(_rad(e)) * math.sin(_rad(L))))
    Etime = q / 15.0 - _fix_angle(RA)
    if Etime > 12:
        Etime -= 24
    elif Etime < -12:
        Etime += 24
    return dec, Etime


def _julian_day(year: int, month: int, day: int) -> float:
    if month <= 2:
        year -= 1
        month += 12
    A = math.floor(year / 100.0)
    B = 2 - A + math.floor(A / 4.0)
    return math.floor(365.25 * (year + 4716)) + math.floor(30.6001 * (month + 1)) + day + B - 1524.5


def _hour_angle(lat: float, dec: float, angle: float) -> float:
    """Returns hour angle in degrees for a given solar angle."""
    cos_ha = (math.sin(_rad(angle)) - math.sin(_rad(lat)) * math.sin(_rad(dec))) / (
        math.cos(_rad(lat)) * math.cos(_rad(dec))
    )
    if cos_ha < -1:
        return 180.0
    if cos_ha > 1:
        return 0.0
    return _deg(math.acos(cos_ha))


def _asr_hour_angle(lat: float, dec: float, shadow_factor: int = 1) -> float:
    """Standard Asr (shadow_factor=1)."""
    target = _deg(math.atan(1.0 / (shadow_factor + math.tan(_rad(abs(lat - dec))))))
    return _hour_angle(lat, dec, target)


def _hours_to_hhmm(h: float) -> str:
    h = h % 24
    total_min = round(h * 60)
    hh = total_min // 60 % 24
    mm = total_min % 60
    return f"{hh:02d}:{mm:02d}"


def compute(date: datetime.date, lat: float = LAT, lon: float = LON, tz: int = TZ_OFFSET) -> dict:
    jd = _julian_day(date.year, date.month, date.day)
    dec, etime = _sun_position(jd)

    # Dhuhr (öğle): solar noon
    dhuhr = 12 + tz - lon / 15.0 - etime

    ha_sunrise = _hour_angle(lat, dec, -0.8333)  # sunrise/sunset
    ha_fajr    = _hour_angle(lat, dec, -FAJR_ANGLE)
    ha_isha    = _hour_angle(lat, dec, -ISHA_ANGLE)
    ha_asr     = _asr_hour_angle(lat, dec, 1)

    fajr    = dhuhr - ha_fajr / 15.0
    sunrise = dhuhr - ha_sunrise / 15.0
    asr     = dhuhr + ha_asr / 15.0
    maghrib = dhuhr + ha_sunrise / 15.0
    isha    = dhuhr + ha_isha / 15.0
    # İmsak: Diyanet'te İmsak = Fajr (aynı açı), bazı kaynaklarda Fajr-10dk
    imsak   = fajr

    return {
        "İmsak":  _hours_to_hhmm(imsak),
        "Güneş":  _hours_to_hhmm(sunrise),
        "Öğle":   _hours_to_hhmm(dhuhr),
        "İkindi": _hours_to_hhmm(asr),
        "Akşam":  _hours_to_hhmm(maghrib),
        "Yatsı":  _hours_to_hhmm(isha),
    }


def as_list(date: datetime.date) -> list[dict]:
    r = compute(date)
    return [{"name": k, "time": v} for k, v in r.items()]

File: frontend/src/prayer-times.ts
// Astronomical prayer time calculation – Diyanet method
// Seydişehir, Turkey: 37.4167°N, 31.85°E, UTC+3
const LAT = 37.4167;
const LON = 31.85;
const TZ = 3;
const FAJR_ANGLE = 18.0;
const ISHA_ANGLE = 17.0;

function rad(d: number) { return (d * Math.PI) / 180; }
function deg(r: number) { return (r * 180) / Math.PI; }
function fixH(h: number) { return h - 24 * Math.floor(h / 24); }

function julianDay(year: number, month: number, day: number): number {
  if (month <= 2) { year -= 1; month += 12; }
  const A = Math.floor(year / 100);
  const B = 2 - A + Math.floor(A / 4);
  return Math.floor(365.25 * (year + 4716)) + Math.floor(30.6001 * (month + 1)) + day + B - 1524.5;
}

function sunPos(jd: number) {
  const D = jd - 2451545.0;
  const g = rad(357.529 + 0.98560028 * D);
  const q = 280.459 + 0.98564736 * D;
  const L = rad(q + 1.915 * Math.sin(g) + 0.02 * Math.sin(2 * g));
  const e = rad(23.439 - 3.6e-7 * D);
  const RA = deg(Math.atan2(Math.cos(e) * Math.sin(L), Math.cos(L))) / 15;
  const eqt = q / 15 - fixH(RA);
  const decl = Math.asin(Math.sin(e) * Math.sin(L));
  return { decl, eqt };
}

function hourAngle(angle: number, decl: number): number {
  const cosH =
    (Math.sin(rad(angle)) - Math.sin(rad(LAT)) * Math.sin(decl)) /
    (Math.cos(rad(LAT)) * Math.cos(decl));
  if (cosH < -1 || cosH > 1) return 0;
  return deg(Math.acos(cosH)) / 15;
}

function fmt(h: number): string {
  h = fixH(h);
  const hh = Math.floor(h);
  const mm = Math.round((h - hh) * 60);
  if (mm === 60) return `${String(hh + 1).padStart(2, '0')}:00`;
  return `${String(hh).padStart(2, '0')}:${String(mm).padStart(2, '0')}`;
}

export function computePrayerTimes(date: Date): Record<string, string> {
  const jd = julianDay(date.getFullYear(), date.getMonth() + 1, date.getDate());
  const { decl, eqt } = sunPos(jd);
  const noon = 12 - LON / 15 - eqt + TZ;
  const srHA = hourAngle(-0.833, decl);
  const sunrise = noon - srHA;
  const sunset = noon + srHA;
  const asrAngle = -deg(Math.atan(1 + Math.tan(Math.abs(rad(LAT) - decl))));
  const asr = noon + hourAngle(asrAngle, decl);
  const fajr = noon - hourAngle(-FAJR_ANGLE, decl);
  const isha = noon + hourAngle(-ISHA_ANGLE, decl);
  return {
    'İmsak': fmt(fajr),
    'Güneş': fmt(sunrise),
    'Öğle': fmt(noon),
    'İkindi': fmt(asr),
    'Akşam': fmt(sunset),
    'Yatsı': fmt(isha),
  };
}

export function getPrayerTimesList(date: Date) {
  const times = computePrayerTimes(date);
  return Object.entries(times).map(([name, time]) => ({ name, time }));
}
